pkg/kube/storage.go

```diff
@@ -0,0 +1,321 @@
+package kube
+
+import (
+	"github.com/skupperproject/skupper/api/types"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Storage interface {
+	Services
+	ConfigMaps
+	Deployments
+	Secrets
+	StatefulSets
+}
+
+type Services interface {
+	GetService(name string) (*corev1.Service, bool, error)
+	DeleteService(name string) error
+	ListServices(options *types.ListFilter) ([]corev1.Service, error)
+	CreateService(svc *corev1.Service) (*corev1.Service, error)
+	UpdateService(svc *corev1.Service) (*corev1.Service, error)
+	IsOwnedService(service *corev1.Service) bool
+}
+
+type ConfigMaps interface {
+	GetConfigMap(name string) (*corev1.ConfigMap, bool, error)
+	DeleteConfigMap(cm string) error
+	ListConfigMaps(options *types.ListFilter) ([]corev1.ConfigMap, error)
+	CreateConfigMap(cm *corev1.ConfigMap) (*corev1.ConfigMap, error)
+	UpdateConfigMap(cm *corev1.ConfigMap) (*corev1.ConfigMap, error)
+	IsOwnedConfigMap(cm *corev1.ConfigMap) bool
+}
+
+type Deployments interface {
+	GetDeployment(name string) (*appsv1.Deployment, bool, error)
+	DeleteDeployment(dep string) error
+	ListDeployments(options *types.ListFilter) ([]appsv1.Deployment, error)
+	CreateDeployment(dep *appsv1.Deployment) (*appsv1.Deployment, error)
+	UpdateDeployment(dep *appsv1.Deployment) (*appsv1.Deployment, error)
+	IsOwnedDeployment(dep *appsv1.Deployment) bool
+}
+
+type Secrets interface {
+	GetSecret(name string) (*corev1.Secret, bool, error)
```

Comment:
I'm not suggesting to change anything, I'm just curious about the following points:
- does the boolean show more information that can not be deducted by the existence of an error? In case of an absence, we still have an error message
- in go we have something equivalent to an `Either<Error, Object>` as the result value of the functions. I wonder if following a structure like `Option<Object>` (a value or an empty value in case of absence) for get functions would be useful.

Reply:
That is a good question. It was based on an existing Services interface used by the controller.

I kept the boolean return parameter to allow the implementation to evaluate the error and eventually
return the error as nil on not founds, to avoid the caller from parsing it itself, as a "not found" error could
be represented differently on each implementation.

But since the subject itself (Secret) would be nil in such case, in spite of an error being returned, I believe it ended
up being just a convenience approach to check whether or not it was found without testing if the subject is not nil.

I think it is simpler to just remove the bool return parameter. Wdyt @nluaces and @grs?

Reply:
I don't have a strong opinion about this; I think we could keep the boolean parameter if it's used in several places.